### ISSUE TITLE
Python virtual environment not used by Studio, completely different Python injected instead

### ISSUE THREAD

**Issue by peter-wimsey** [aws-samples/sagemaker-studio-custom-image-samples#18]
Somewhat similar to #9

I have a Docker image with a virtual environment and dependencies installed by pip:

```bash
# Dockerfile
RUN python -m venv /opt/venv \
    && . /opt/venv/bin/activate \
    && pip install --upgrade pip wheel \
    && pip install --requirement requirements.txt \
    && pip cache purge \
    && python -m ipykernel install
```

The image has a kernelspec in the right place:

```bash
# in the image terminal
$ cat /usr/local/share/jupyter/kernels/python3/kernel.json
```
```json
{
 "argv": [
  "/opt/venv/bin/python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3 (ipykernel)",
 "language": "python",
 "metadata": {
  "debugger": true
 }
}
```

**However**, in Studio the virtual environment is not used, instead I get:

```python
# in Studio with image kernel running
import sys
sys.exectuable
# '/opt/.sagemakerinternal/conda/bin/python'
```

which is some Python which isn't even in the image:

```bash
# in image terminal
ls -lh /opt/.sagemakerinternal/conda/bin/python3
# lrwxrwxrwx 1 nobody nogroup 9 Oct 12 17:38 /opt/.sagemakerinternal/conda/bin/python3 -> python3.7
ls -lh /opt/.sagemakerinternal/conda/bin/python3.7
# -rwxr-xr-x 1 nobody nogroup 13M Oct 12 17:38 /opt/.sagemakerinternal/conda/bin/python3.7
```

I was expecting Studio to look through the usual kernelspec locations and find the one matching the config for the attached image (which is set to a kernel called "python3").

What do I have to do for Studio to use the virtual environment as a kernel?

**Comment by peter-wimsey** [comment]
Seems that SageMaker Studio **does not** look for kernel specs in the system locations listed here:

https://jupyter-client.readthedocs.io/en/5.2.3/kernels.html#kernel-specs

That is, it does not find the kernel I installed in `/usr/local/share/jupyter/kernels`.

Instead it looks for the kernel spec in `{sys.prefix}/share/jupyter/kernels` so 

```bash
python -m ipykernel install --sys-prefix   # good, works in SageMaker Studio
```
instead of

```bash
python -m ipykernel install  # bad, does not work in SageMaker Studio
```

solved my problem.

**Comment by wdwatkins** [comment]
Thank you for posting this, I ran into a similar issue trying to get rocker R images (`rocker/binder` specifically) to work in Studio.  It turns out that the binder image installs the kernel to the python venv path: https://github.com/rocker-org/rocker-versioned2/blob/master/scripts/install_binder.sh#L29.  Simply adding another `R -e "IRkernel::installspec()"` to my Dockerfile fixed the kernel not being found error.

In case anyone else finds this, this Dockerfile worked for me (also addressed the seemingly very limited GID range allowed, which I did not find documented anywhere):

```
FROM rocker/binder:4.1.2

USER root

RUN usermod -g users rstudio
RUN usermod -l sagemaker-user rstudio

USER sagemaker-user

RUN pip3 install --user boto3 sagemaker
RUN R -e "IRkernel::installspec()"
```
